src/main/java/com/spotify/confidence/EventSenderEngineImpl.java

```diff
@@ -1,162 +1,171 @@
+        .setPayload(
+            Struct.newBuilder()
+                .putAllFields(message.orElse(ConfidenceValue.Struct.EMPTY).asProtoMap())
+                .putFields(CONTEXT, context.toProto()));
   }
 
-  private void flush() {
-    // Cancel the existing scheduled task if it exists
-    if (pendingFlush != null && !pendingFlush.isDone()) {
-      pendingFlush.cancel(false);
+  @Override
+  public void send(
+      String name, ConfidenceValue.Struct context, Optional<ConfidenceValue.Struct> message) {
+    if (intakeClosed) {
+      log.warn("EventSenderEngine is closed, dropping event {}", name);
+      return;
     }
-    eventStorage.createBatch();
-    uploadQueue.add(UPLOAD_SIG);
-  }
-
-  class UploadPoller implements Runnable {
-    @Override
-    public void run() {
-      while (true) {
-        try {
-          final String signal = uploadQueue.take();
-          final List<EventBatch> batches = List.copyOf(eventStorage.getBatches());
-          for (EventBatch batch : batches) {
-            final boolean uploadSuccessful = eventUploader.upload(batch).get();
-            if (uploadSuccessful) {
-              eventStorage.deleteBatch(batch.id());
-            }
-          }
-          if (signal.equals(SHUTDOWN_UPLOAD)) {
-            shutdownQueue.add(SHUTDOWN_UPLOAD_COMPLETED);
-          }
-        } catch (InterruptedException | ExecutionException e) {
-          throw new RuntimeException(e);
-        }
-      }
+    final Event event = event(name, context, message).setEventTime(clock.getTimestamp()).build();
+    if (estimatedMemoryConsumption.get() + event.getSerializedSize() > maxMemoryConsumption) {
+      log.warn("EventSenderEngine is overloaded, dropping event {}", name);
+      return;
     }
+    sendQueue.add(event);
+    estimatedMemoryConsumption.addAndGet(event.getSerializedSize());
+    LockSupport.unpark(pollingThread);
   }
 
-  @Override
-  public void send(
-      String name, ConfidenceValue.Struct context, Optional<ConfidenceValue.Struct> message) {
-    if (!isStopped) {
-      log.trace("Sending event: {}", name);
-      writeQueue.add(
-          new Event(
-              EVENT_NAME_PREFIX + name,
-              message.orElse(ConfidenceValue.Struct.EMPTY),
-              context,
-              clock.currentTimeSeconds()));
+  private void pollLoop() {
+    Instant latestFlushTime = Instant.now();
+    ArrayList<com.spotify.confidence.events.v1.Event> events = new ArrayList<>();
+    while (true) {
+
+      final var event = sendQueue.poll();
+      if (event != null) {
+        events.add(event);
+      } else {
+        if (intakeClosed) break;
+        LockSupport.parkUntil(Instant.now().plus(maxFlushInterval).toEpochMilli());
+      }
+      final boolean passedMaxFlushInterval =
+          !maxFlushInterval.isZero()
+              && Duration.between(latestFlushTime, Instant.now()).compareTo(maxFlushInterval) > 0;
+      if (events.size() == maxBatchSize || passedMaxFlushInterval) {
```

Comment:
if I understand the park correctly, if the current event is null, we wait until the `maxFlushInterval` and then here we check if we are passed that time (which in this case will be true) and then we upload events. should we check here as well if events are not empty?